Add 'alias mode' support for show commands

paavaanan opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "closes #xxxx",
"fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
issue when the PR is merged

Please provide the following information:
-->

**- What I did**

Added portalias support for following commands.  To enable alias output in CLI execute "sudo config interface_naming_mode alias" command and use the below commands.

- show ip route
- show ipv6 route
- show interface transceiver eeprom
- show arp 
- show lldp neighbors
- show vlan brief
- show vlan config

**- How I did it**

Implemented a seperate run_command_in_alias_mode function  to change the results of linux commands from default interface name to vendor specific name. This function read each line of the output and replace the equivalent vendor names. To speed up the result for mutli-line output ( show ip route/ipv6 route/arp)  we cache the alias_names from port_config.ini at init. This results in quick substitution.

**- How to verify it**
Execute the commands as described above

paavaanan commented on show/main.py:
```diff
@@ -28,6 +29,12 @@ class Config(object):
     def __init__(self):
         self.path = os.getcwd()
         self.aliases = {}
+        global port_dict
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -129,6 +136,58 @@ def run_command(command, display_cmd=False):
+def interface_name_to_alias(interface_name):
+    """Return alias interface name if default
+       name is given as argument
+
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -129,6 +136,58 @@ def run_command(command, display_cmd=False):
+        for port_name in port_dict.keys():
+            if interface_name == port_name:
+                return port_dict[port_name]['alias']
+        print "Invalid interface {}".format(interface_name)
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -129,6 +136,58 @@ def run_command(command, display_cmd=False):
+
+
+def run_command_in_alias_mode(command, linux_command=False):
+    """Replace command output results from default
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -129,6 +136,58 @@ def run_command(command, display_cmd=False):
+             break
+        if output:
+            if linux_command:
+                result = re.findall('Ethernet\w+', output)
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -129,6 +136,58 @@ def run_command(command, display_cmd=False):
+                if interface_name != "":
+                    output = output.replace(interface_name,
+                                    interface_name_to_alias(interface_name))
+                print output.rstrip('\n')
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -129,6 +136,58 @@ def run_command(command, display_cmd=False):
+            else:
+                """Framework need to be designed for other tabulated
+                   commands
+                """
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -955,7 +1034,11 @@ def tablelize(keys, data):
                 r.append(data[k]['vlanid'])
-                r.append(m)
+                if get_interface_mode() == "alias":
+                    alias =  interface_name_to_alias(m)
```
Addressed in the new commit

paavaanan commented:
In this new pull request implemented portalias feature for sonic-specific commands mentioned below:
- show pfc counters
- show queue counters
- show interface status
- show interface description
- show interface transceiver lpmode
- show interface transceiver presence
- show interface transceiver eeprom
- show lldp table
- show mac

As discussed implemented an Alias class for most of the portalias related functions. Most the above commands are in tabulated form. In function run_command_in_alias_mode else part handles above commands. Above command reusults consist of 3 parts. Header, Underline, and tabulate data. These are splitted into different categories and handled accordingly. 


```
Header------------> Adjusted spacing with the help of vendor_name_length variable
Underline---------> Introduced index variable to pick the column to be adjusted.  
Data--------------> String replace default name to alias name.
```

jleveque commented on show/main.py:
```diff
@@ -38,6 +40,45 @@ def read_config(self, filename):
+                if interface_name == port_name:
+                    return self.port_dict[port_name]['alias']
+        click.echo("Invalid interface {}".format(interface_name))
+
```
Remove blank line 67. Consider adding blank line between 65 and 66.

paavaanan replied:
👍

jleveque commented on show/main.py:
```diff
@@ -38,6 +40,45 @@ def read_config(self, filename):
+                if interface_alias == self.port_dict[port_name]['alias']:
+                    return port_name
+        click.echo("Invalid interface {}".format(interface_alias))
+
```
Remove blank line 78. Consider adding blank line between 76 and 77.

paavaanan replied:
👍

jleveque commented on show/main.py:
```diff
@@ -129,6 +170,112 @@ def run_command(command, display_cmd=False):
+                output = output.lstrip()
+                index = 0
+
+                if command_type == 1:
```
`command_type` values as simple integers are not clearly defined (what does "1" mean? etc.). This makes them confusing. Also, if we have one "command_type" for each command, this will become unwieldy. It would be better if you could key off the command string itself, and not rely on an integer which represents the command.

paavaanan replied:
This is good one. Initially I thougt of using enums,  but due to lack of package I ruled out the option. As suggested I used the command strings directly in the function and included in the latest commit.

paavaanan commented on show/main.py:
```diff
@@ -38,6 +40,45 @@ def read_config(self, filename):
             pass
 
 
+class Alias(object):
```
Addressed in the new commit...

paavaanan commented on show/main.py:
```diff
@@ -38,6 +40,45 @@ def read_config(self, filename):
 
 
+class Alias(object):
+    """Object to Alias conversion"""
```
Addressed in the new commit

paavaanan commented on show/main.py:
```diff
@@ -38,6 +40,45 @@ def read_config(self, filename):
+    """Object to Alias conversion"""
+
+    def __init__(self):
+        self.vendor_name_length = 0
```
Addressed in the new commit

jleveque commented on show/main.py:
```diff
@@ -129,6 +170,110 @@ def run_command(command, display_cmd=False):
+        if output == '' and process.poll() is not None:
+            break
+        if output:
+            if linux_command:
```
Is there a way to make Linux commands the default (i.e., "else") case here, such that if the command doesn't match any of the if/elif cases, it is treated as a Linux command? This would eliminate the need for the `linux_command` parameter.

paavaanan replied:
I like the idea. But, without the linux_command it triggers lot of confusion in replacing interfaces from defautl to alias Currently we are using different substitution methods for linux commands and sonic utilities. Without  this differentiation we are forced to club the logic. This is an overhead to take care all possibilities.

jleveque replied:
Could you explain what you mean by "we are forced to club the logic"? My thought is that if the command doesn't match any of the specific `if command == ...` clauses, it would fall through into a "default" case, and this case would handle all Linux commands.

paavaanan replied:
Joe, the linux/sonic commands use 2 different logics in replacing the interface names. I tested the else style of approach what you suggested. I moved the sonic_utitilies replacement logic inside a generic function "print_output_in_alias_mode"  and included in most of the  "if" statements.  
```
if command == "portstat":
               """Show interface counters"""
               if output.startswith("IFACE"):
                   output = output.replace("IFACE", "IFACE".rjust(
                                           _interface.vendor_name_max_length))
               print_output_in_alias_mode(output, index)<----

            elif command == "pfcstat":
               """Show pfc counters"""
               if output.startswith("Port Tx"):
                   output = output.replace("Port Tx", "Port Tx".rjust(
                                           _interface.vendor_name_max_length))

               elif output.startswith("Port Rx"):
                   output = output.replace("Port Rx", "Port Rx".rjust(
                                           _interface.vendor_name_max_length))
               print_output_in_alias_mode(output, index)<----
```
Since, sonic_utitilies replacement logic is safely binded inside the if statements now it is free to introduce else for linux_utitlites commands. This approach works fine. 

I agree with checking alias mode inside the run_command. Will raise pull request post testing.

paavaanan commented on show/main.py:
```diff
@@ -38,6 +40,45 @@ def read_config(self, filename):
+        click.echo("Invalid interface {}".format(interface_name))
+        raise click.Abort()
+
+    def interface_alias_to_name(self, interface_alias):
```
Addressed in the new commit..

paavaanan commented on show/main.py:
```diff
@@ -129,6 +170,110 @@ def run_command(command, display_cmd=False):
+
+
+# Global Alias object
+_alias = InterfaceAliasConverter()
```
Addressed in the new commit..

paavaanan commented on show/main.py:
```diff
@@ -129,6 +170,110 @@ def run_command(command, display_cmd=False):
+                    elif output.startswith("Port Rx"):
+                        output = output.replace("Port Rx", "Port Rx".rjust(
+                                                _alias.vendor_name_max_length))
+                elif (command == "sudo sfputil show eeprom") or(
```
Addressed in the new commit..

paavaanan commented on show/main.py:
```diff
@@ -129,6 +173,148 @@ def run_command(command, display_cmd=False):
+        word[index] = underline
+        output = '  ' .join(word)
+
+    # Replace default interface name to vendor-name
```
```
# Global class instance for SONiC interface name to alias conversion
convert_interface = InterfaceAliasConverter()
```
Renamed as convert_interface so, if name_to_alias / alias_to_name functions are used along with this variable name it will be more meaningful..

```
if get_interface_mode() == "alias":
1174                     alias = convert_interface.name_to_alias(m)
1175                     r.append(alias)
1176                 else:
1177                     r.append(m)
```

paavaanan commented on show/main.py:
```diff
@@ -129,6 +173,148 @@ def run_command(command, display_cmd=False):
+
+
+# Global Alias object
+_interface = InterfaceAliasConverter()
```
```
# Global class instance for SONiC interface name to alias conversion
convert_interface = InterfaceAliasConverter()
```
Renamed as convert_interface so, if name_to_alias / alias_to_name functions are used along with this variable name it will be more meaningful..

``` 
               if get_interface_mode() == "alias":
                    alias = convert_interface.name_to_alias(m)
                    r.append(alias)
                else:
                    r.append(m)

```

jleveque replied:
"convert_interface" sounds like a function, not an object. It's hard not to get too verbose, but I think it should be named something like "iface_alias_converter".

paavaanan replied:
Addressed as discussed in new commit..

paavaanan commented on show/main.py:
```diff
@@ -116,6 +156,10 @@ def run_command(command, display_cmd=False):
     if display_cmd:
         click.echo(click.style("Command: ", fg='cyan') + click.style(command, fg='green'))
 
+    if get_interface_mode() == "alias" and not command.startswith("intfutil"):
```
Addressed as above

jleveque commented on show/main.py:
```diff
@@ -38,6 +39,45 @@ def read_config(self, filename):
+    """Class which handles conversion between interface name and alias"""
+
+    def __init__(self):
+        self.vendor_name_max_length = 0
```
Please rename to "alias_max_length"

paavaanan replied:
Addressed in new commit!

jleveque commented on show/main.py:
```diff
@@ -207,6 +402,10 @@ def alias(interfacename):
     body = []
 
     if interfacename is not None:
+
```
Remove blank line

paavaanan replied:
👍

paavaanan commented on show/main.py:
```diff
@@ -38,6 +39,45 @@ def read_config(self, filename):
+        raise click.Abort()
+
+    def alias_to_name(self, interface_alias):
+        """Return default interface name if alias name is given as argument
```
👍

paavaanan commented on show/main.py:
```diff
@@ -38,6 +39,45 @@ def read_config(self, filename):
+                    self.port_dict[port_name]['alias'])
+
+    def name_to_alias(self, interface_name):
+        """Return alias interface name if default
```
👍

jleveque commented on show/main.py:
```diff
@@ -129,6 +176,150 @@ def run_command(command, display_cmd=False):
 
+
+def get_interface_mode():
+    mode = os.getenv('SONIC_CLI_IFACE_MODE')
```
This solution is currently not working because you are using `os.getenv()` to check the `SONIC_CLI_IFACE_MODE` variable. However, this variable is not available to subprocesses (i.e., the "show" or "config" commands) because it is not exported in .bashrc.

You will need to modify the "config" command and add the `export` keyword when writing the `SONIC_CLI_IFACE_MODE` variable to the .bashrc file.

paavaanan replied:
SONIC_CLI_IFACE_MODE env variable is working fine acorss sessions/users. This is achieved by exporting the varaiable in  /etc/environment and sudoers file of the switch. 

Already raised a pull request  on this: 
https://github.com/Azure/sonic-buildimage/pull/1751/

We store variable in .bashrc (each user) for persistence. We finalised sudoers/environment approach because there is no direct way to change the  the variable of parent from the forked child bash instance. This already we agreed upon and the pull-request is raised on that design.

jleveque replied:
I was under the assumption we abandoned the sudoers/environment approach (Azure/sonic-buildimage#1751) in favor of asking the user to log out and log back in again.

paavaanan replied:
We can safely discard that pull request. SONIC_CLI_IFACE_MODE variable changes are applied to bash environment when user logout and logback in to the new shell. So, there is no any necessity for /etc/environment and /etc/sudoers file to change the values in realtime. I raised a new pull-request in below with the above changes:

https://github.com/Azure/sonic-utilities/pull/321

jleveque replied:
👍

jleveque commented:
Is there a reason you didn't add alias-to-name conversion for the "show interfaces transceiver ..." commands? If not, please add it.

Also, please fix the merge conflict (it's only whitespace-related).

paavaanan commented:
This is not added due to the lack of xvcrd implemetation in Dell boxes. Since it is added recently, I can able to test with the new change. Issue addresed in new commit.

Addressed conflicts and added "show interface naming_mode" command support in new commit..

jleveque commented on show/main.py:
```diff
@@ -129,6 +176,156 @@ def run_command(command, display_cmd=False):
+                                iface_alias_converter.alias_max_length))
+                print_output_in_alias_mode(output, index)
+
+            elif command == "sudo sfputil show eeprom":
```
This should be `command.startswith("sudo sfputil show eeprom")` because the command can take flags and an optional interface argument.

You also need to add code to translate the optional interface name from vendor alias to SONiC name in the "show interfaces transceiver ..." commands on lines 506-548 like you did in the other commands (see lines 560-562 for reference).

paavaanan replied:
Addressed and updated in new commit.. 👍

jleveque commented:
@qiluo-msft: The root cause was due to the lack of a "PORT" table in ConfigDB on the virtual switch. The "PORT" table will be populated by this PR: https://github.com/Azure/sonic-buildimage/pull/2169

nikos-github commented:
@paavaanan @jleveque Your commit is breaking 'show interface status', 'show platform summary' and other cmds. Alias is an optional field. See https://github.com/Azure/sonic-utilities/issues/403
